Content/Items/Vitric/Weapons.CoachGunUpgrade.cs

```diff
@@ -0,0 +1,686 @@
+        {
+            trail = trail ?? new Trail(Main.instance.GraphicsDevice, 40, new TriangularTip(40 * 3.5f), factor => width, factor =>
+            {
+                return new Color(255, 100 + (int)(100 * (float)Math.Sin(TimeFade * 3.14f)), 65) * (float)Math.Sin(TimeFade * 3.14f) * 0.5f;
+            });
+
+            trail.Positions = cache.ToArray();
+            trail.NextPosition = cache[39];
+        }
+
+        public void DrawPrimitives()
+        {
+            if (Projectile.ai[1] == 1f)
+                return;
+
+            Effect effect = Filters.Scene["CeirosRing"].GetShader().Shader;
+
+            Matrix world = Matrix.CreateTranslation(-Main.screenPosition.Vec3());   
+            Matrix view = Main.GameViewMatrix.ZoomMatrix;
+            Matrix projection = Matrix.CreateOrthographicOffCenter(0, Main.screenWidth, Main.screenHeight, 0, -1, 1);
+
+            effect.Parameters["time"].SetValue(Projectile.timeLeft * 0.01f);
+            effect.Parameters["repeats"].SetValue(6);
+            effect.Parameters["transformMatrix"].SetValue(world * view * projection);
+            effect.Parameters["sampleTexture"].SetValue(ModContent.Request<Texture2D>("StarlightRiver/Assets/EnergyTrail").Value);
+
+            trail?.Render(effect);
+
+            effect.Parameters["sampleTexture"].SetValue(ModContent.Request<Texture2D>("StarlightRiver/Assets/FireTrail").Value);
+
+            trail?.Render(effect);
+        }
+    }
+
+    public class VitricBombDust : ModDust
+    {
+        public override string Texture => AssetDirectory.Dust + Name;
+
+        public override void OnSpawn(Dust dust)
+        {
+            dust.noLight = false;
+        }
+
+        public override bool Update(Dust dust)
+        {
+            dust.position += dust.velocity;
+            dust.rotation += 0.12f;
+            dust.scale *= 0.98f;
+            if (dust.scale <= 0.2)
+                dust.active = false;
+
+            if (!dust.noGravity)
+                dust.velocity.Y += 0.15f;
+
+            return false;
+        }
+    }
+
+    class SwelteredDeBuff : SmartBuff
+    {
+        public override string Texture => AssetDirectory.Debug;
+
+        public SwelteredDeBuff() : base("Sweltered", "Damage taken increased by 35%", false) { }
+
+        public override void Update(NPC npc, ref int buffIndex)
+        {
+            npc.GetGlobalNPC<SwelteredGlobalNPC>().Sweltered = true;
+
+            Vector2 vel = new Vector2(0, -1).RotatedByRandom(0.5f) * 0.4f;
+            if (Main.rand.NextBool(4))
+                Dust.NewDust(npc.position, npc.width, npc.height, ModContent.DustType<CoachSmoke>(), vel.X, vel.Y, 0, new Color(60, 55, 50) * 0.5f, Main.rand.NextFloat(0.5f, 1));
+
+            if (Main.rand.NextBool(2))
+                Dust.NewDust(npc.position, npc.width, npc.height, ModContent.DustType<Dusts.Glow>(), 0, 0, 0,Color.DarkOrange, 0.6f);
+        }
+    }
+    public class SwelteredGlobalNPC : GlobalNPC
```

Comment:
See the StarlightNPC Delegation pattern. It should most likely be used here instead of a GlobalNPC.